### PR TITLE
apiserver/metricsdebug: Added a new metricsdebug facade and a GetMetr…

### PR DESCRIPTION
**Pull request by mattyw**
…ics implementation

### CODE REVIEW

File: apiserver/params/params.go [hunk 1]
```diff
@@ -804,3 +804,18 @@ type BundleChangesChange struct {
...
+type MetricsResults struct {
+	Results []MetricsResult `json:"results"`
+}
+
+type MetricsResult struct {
+	Metric []MetricResult `json:"metric,ommitempty"`
+	Error  *Error         `json:"error,omitempty"`
+}
+
+type MetricResult struct {
```
Comment (alesstimec) [inline apiserver/params/params.go]: can we talk about this at the standup?

File: apiserver/metricsdebug/metricsdebug.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+var _ MetricsDebug = (*MetricsDebugAPI)(nil)
+
+// NewMetricsDebugAPI creates a new API endpoint for calling metrics debug functions.
+func NewMetricsDebugAPI(
+	st *state.State,
+	resources *common.Resources,
+	authorizer common.Authorizer,
+) (*MetricsDebugAPI, error) {
+	if !(authorizer.AuthMachineAgent() && authorizer.AuthEnvironManager()) {
+		// TODO (mattyw) Needs to be different.
```
Comment (cmars) [inline apiserver/metricsdebug/metricsdebug.go]: What is it supposed to be?

File: apiserver/metricsdebug/metricsdebug.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	}
+	if len(args.Entities) == 0 {
+		return results, nil
+	}
+	canAccess, err := api.accessEnviron()
+	if err != nil {
+		return results, err
+	}
+	for i, arg := range args.Entities {
+		tag, err := names.ParseEnvironTag(arg.Tag)
```
Comment (cmars) [inline apiserver/metricsdebug/metricsdebug.go]: I think we'd want service and unit tags here. Is this just a first incremental step?

**Comment by cmars** [comment]
Some questions.

**Comment by cmars** [comment]
LGTM as long as there are followups to the above.

File: api/metricsdebug/client_test.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+type metricsdebugSuite struct {
+	jujutesting.JujuConnSuite
+
+	manager *metricsdebug.Client
+}
+
+var _ = gc.Suite(&metricsdebugSuite{})
+
+func (s *metricsdebugSuite) SetUpTest(c *gc.C) {
+	s.JujuConnSuite.SetUpTest(c)
```
Comment (cmars) [inline api/metricsdebug/client_test.go]: Would it be a good idea to split this test suite into two -- one with mocks and one with the connsuite setup? That way we don't waste time setting up and tearing down the connsuite stuff for tests that don't need them.

File: apiserver/metricsdebug/metricsdebug_test.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+func (s *metricsDebugSuite) SetUpTest(c *gc.C) {
+	s.JujuConnSuite.SetUpTest(c)
+	s.authorizer = apiservertesting.FakeAuthorizer{
+		Tag: s.AdminUserTag(c),
+	}
+	debug, err := metricsdebug.NewMetricsDebugAPI(s.State, nil, s.authorizer)
+	c.Assert(err, jc.ErrorIsNil)
+	s.metricsdebug = debug
+	/*
```
Comment (cmars) [inline apiserver/metricsdebug/metricsdebug_test.go]: Remove this commented out stuff?

**Comment by cmars** [comment]
Looking pretty good. In addition to the suggestions above, I'd like godocs on exported symbols.

File: api/metricsdebug/client_test.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+	"github.com/juju/juju/api/metricsdebug"
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/params"
+	jujutesting "github.com/juju/juju/juju/testing"
+	"github.com/juju/juju/testing/factory"
+)
+
+type metricsdebugSuite struct {
+	jujutesting.JujuConnSuite
```
Comment (alesstimec) [inline api/metricsdebug/client_test.go]: how about including a Stub

File: apiserver/metricsdebug/metricsdebug.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+		return nil, common.ErrPerm
+	}
+
+	return &MetricsDebugAPI{
+		state: st,
+	}, nil
+}
+
+func (api *MetricsDebugAPI) unitOrService(entity string) (string, error) {
+	if _, err := api.state.Unit(entity); err == nil {
```
Comment (alesstimec) [inline apiserver/metricsdebug/metricsdebug.go]: Perhaps you could return an enum type here instead of string..

File: apiserver/params/params.go [hunk 8]
```diff
@@ -804,3 +804,28 @@ type BundleChangesChange struct {
 	// before this change is applied.
 	Requires []string `json:"requires"`
 }
+
+type MetricsResults struct {
```
Comment (alesstimec) [inline apiserver/params/params.go]: godoc, please

File: apiserver/params/params.go [hunk 9]
```diff
@@ -804,3 +804,28 @@ type BundleChangesChange struct {
...
+func (m *MetricsResults) OneError() error {
+	for _, r := range m.Results {
+		if err := r.Error; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type MetricsResult struct {
```
Comment (alesstimec) [inline apiserver/params/params.go]: godoc, please

File: apiserver/params/params.go [hunk 10]
```diff
@@ -804,3 +804,28 @@ type BundleChangesChange struct {
...
+	Metrics []MetricResult `json:"metrics,omitempty"`
+	Error   *Error         `json:"error,omitempty"`
+}
+
+type MetricResult struct {
```
Comment (alesstimec) [inline apiserver/params/params.go]: godoc please

**Comment by alesstimec** [comment]
LGTM with a few nitpicks

File: apiserver/metricsdebug/metricsdebug.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
+// Copyright 2015 Canonical Ltd.
```
Comment (mattyw) [inline apiserver/metricsdebug/metricsdebug.go]: 2016

File: apiserver/metricsadder/metricsadder_test.go [hunk 12]
```diff
@@ -101,7 +101,7 @@ func (s *metricsAdderSuite) TestAddMetricsBatch(c *gc.C) {
 	metrics := []params.Metric{{Key: "pings", Value: "5", Time: time.Now().UTC()}}
 	uuid := utils.MustNewUUID().String()
 
-	result, err := s.adder.AddMetricBatches(params.MetricBatchParams{
+	result, err := s.adder.AddAllMetricBatches(params.MetricBatchParams{
```
Comment (mattyw) [inline apiserver/metricsadder/metricsadder_test.go]: wrong

File: apiserver/metricsadder/metricsadder_test.go [hunk 13]
```diff
@@ -133,7 +133,7 @@ func (s *metricsAdderSuite) TestAddMetricsBatchNoCharmURL(c *gc.C) {
 	metrics := []params.Metric{{Key: "pings", Value: "5", Time: time.Now().UTC()}}
 	uuid := utils.MustNewUUID().String()
 
-	result, err := s.adder.AddMetricBatches(params.MetricBatchParams{
```
Comment (mattyw) [inline apiserver/metricsadder/metricsadder_test.go]: wrong

File: apiserver/metricsdebug/metricsdebug_test.go [hunk 14]
```diff
@@ -0,0 +1,154 @@
+// Copyright 2014 Canonical Ltd.
```
Comment (mattyw) [inline apiserver/metricsdebug/metricsdebug_test.go]: 2016

File: apiserver/metricsdebug/package_test.go [hunk 15]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2014 Canonical Ltd.
```
Comment (mattyw) [inline apiserver/metricsdebug/package_test.go]: 2016

File: apiserver/params/params.go (hunk 8 above)
Comment (mattyw) [inline apiserver/params/params.go]: doc

File: state/metrics_test.go [hunk 16]
```diff
@@ -248,7 +248,7 @@ func (s *MetricSuite) TestCleanupMetricsIgnoreNotSent(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 }
 
-func (s *MetricSuite) TestMetricBatches(c *gc.C) {
+func (s *MetricSuite) TestAllMetricBatches(c *gc.C) {
```
Comment (mattyw) [inline state/metrics_test.go]: wrong

File: state/metrics_test.go [hunk 17]
```diff
@@ -333,7 +423,7 @@ func (s *MetricSuite) TestCountMetrics(c *gc.C) {
 	c.Assert(unsent+sent, gc.Equals, 3)
 }
 
-func (s *MetricSuite) TestSetMetricBatchesSent(c *gc.C) {
+func (s *MetricSuite) TestSetAllMetricBatchesSent(c *gc.C) {
```
Comment (mattyw) [inline state/metrics_test.go]: wrong

File: api/metricsdebug/client_test.go (hunk 4 above)
Comment (mattyw) [inline api/metricsdebug/client_test.go]: @cmars it's still needed for some of the stup in the apiserver. even though we end up mocking part of it

Comment (cmars) [inline api/metricsdebug/client_test.go]: Fair enough. Feel free to ignore then.

File: apiserver/metricsdebug/metricsdebug.go [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/state"
+)
+
+var (
+	logger = loggo.GetLogger("juju.apiserver.metricsdebug")
+)
+
+func init() {
+	common.RegisterStandardFacade("MetricsDebug", 0, NewMetricsDebugAPI)
```
Comment (cmars) [inline apiserver/metricsdebug/metricsdebug.go]: I think new facade versions start numbering at 1.

Comment (mattyw) [inline apiserver/metricsdebug/metricsdebug.go]: @cmars done

**Pull request closed without merging**